pkg/kubelet/network/kubenet/kubenet_linux.go

```diff
@@ -347,14 +321,20 @@ func (plugin *kubenetNetworkPlugin) SetUpPod(namespace string, name string, id k
 	shaper := plugin.shaper()
 
 	if egress != nil || ingress != nil {
-		ipAddr := plugin.podIPs[id]
-		if err := shaper.ReconcileCIDR(fmt.Sprintf("%s/32", ipAddr), egress, ingress); err != nil {
+		if err := shaper.ReconcileCIDR(fmt.Sprintf("%s/32", ip4.String()), egress, ingress); err != nil {
 			return fmt.Errorf("Failed to add pod to shaper: %v", err)
 		}
 	}
 
-	plugin.syncHostportsRules()
-	return nil
+	plugin.podIPs[id] = ip4.String()
+
+	// Open any hostports the pod's containers want
+	runningPods, err := plugin.getRunningPods()
+	if err == nil {
+		err = plugin.hostportHandler.OpenPodHostportsAndSync(pod, BridgeName, runningPods)
```

Comment:
> Say if the CNI ADD succeeded, but port is in use. SetUpPod will end up setting up the veth but fail at opening hostport. IIUC, kubelet will just kill infra container without calling TearDownPod (it assumes SetUpPod won't be partially successful). Then this Pod will stuck, because CNI ADD will always return failure since eth0 has been setup.

You are correct that this PR will exacerbate the situation a bit, so I'll try to fix that.  However, it looks like all the network plugins have problems with this, since they don't necessarily clean up after themselves on errors in SetUpPod().  And since (at least) dockertools/manager.go never calls TearDownPod() when setup fails, I think there are already cases where failures will leave the pod network configured...  I'll try to figure something out here.